src/soc/ibm/power9/mvpd.c

```diff
@@ -199,13 +201,15 @@ static bool eeprom_extract_kwd(uint64_t offset, uint8_t index,
 	return false;
 }
 
-/* Builds MVPD partition for a single processor (64 KiB per chip) or returns an
+/* Builds MVPD partition for a single processor (up to 64 KiB) or returns an
  * already built one */
-static const uint8_t *mvpd_get(void)
+static const uint8_t *mvpd_get(uint8_t cpu)
 {
-	enum { SECTION_SIZE = 64 * KiB };
+	/* Actual size of MVPD is a bit less than 42 KiB while maximum is 64
+	 * KiB, save some memory */
+	enum { MVPD_SIZE = 42 * KiB };
```

Comment:
We can make region reserved for romstage bigger, but that wouldn't help with lag. I guess MVPD in PNOR was created specifically because of this. Some work on asynchronous loading of stages was made on upstream coreboot, I haven't dig into that yet, but maybe we could use it somehow.

We can make a separate section for MVPD in memlayout.ld, but we would probably have to use lower 2MB of cache which would require changes in bootblock, or make `CBFS_CACHE` smaller. `CBFS_CACHE` must be big enough for whole CBFS, MEMD and MVPD, if we also want to write it back to PNOR. Both options seems doable, I can try tomorrow.

IIRC cbmem is initialized at the end of romstage so this wouldn't give us more space here, but at least that is a way of passing data to ramstage, if everything else fails.

I also planned to cache whole CBFS in L3 cache, but this may be too much to fit. Reading from EEPROM is definitely slower so it is more important. Another thing is SPD, with Hostboot it is added to device tree. Can you think of any other things worth caching?

Reply:
> IIRC cbmem is initialized at the end of romstage so this wouldn't give us more space here, [...]

Ah, of course... It stores things in RAM.

> Can you think of any other things worth caching?

No, nothing else comes to mind.

Reply:
https://github.com/Dasharo/coreboot/pull/82